cmd/root.go

```diff
@@ -0,0 +1,227 @@
+package cmd
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/push"
+	"github.com/spf13/cobra"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/util/homedir"
+)
+
+var (
+	rootCmd = &cobra.Command{
+		Use:   "zombie-detector",
+		Short: "zombie-detector detects longly undeleted kubernetes resources",
+		RunE:  rootMain,
+	}
+)
+
+const (
+	errorFaildtoList      string = "the server could not find the requested resource"
+	errorMethodNotAllowed string = "the server does not allow this method on the requested resource"
+)
+
+var inClusterFlag bool
+var thresholdFlag string
+var pushgatewayEndpointFlag string
+
+func init() {
+	rootCmd.AddCommand(versionCmd)
+	rootCmd.Flags().BoolVar(&inClusterFlag, "incluster", true, "execute in cluster or not")
+	rootCmd.Flags().StringVar(&thresholdFlag, "threshold", "24h", "threshold of detection")
+	rootCmd.Flags().StringVar(&pushgatewayEndpointFlag, "pushgateway", "", "URL of Pushgateway's endpoint")
+
+}
+
+func Execute() {
+	if err := rootCmd.Execute(); err != nil {
+		log.Fatal(err)
+	}
+}
+
+func setupClusterConfig(inCluster bool) (*rest.Config, error) {
+	if inCluster {
+		config, err := rest.InClusterConfig()
+		if err != nil {
+			return nil, err
+		}
+		return config, nil
+	}
+	var kubeconfig string
+	if home := homedir.HomeDir(); home != "" {
+		kubeconfig = filepath.Join(home, ".kube", "config")
+	} else {
+		return nil, errors.New("Faild to load kubeconfig")
+	}
+	return clientcmd.BuildConfigFromFlags("", kubeconfig)
+}
+
+func getAllResources(ctx context.Context, config *rest.Config) ([]unstructured.Unstructured, error) {
+	o, err := discovery.NewDiscoveryClientForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+	dynamicClient, err := dynamic.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+	serverResources, err := o.ServerPreferredResources()
+	if err != nil {
+		return nil, err
+	}
+	resources := make([]unstructured.Unstructured, 0)
+	for _, resList := range serverResources {
+		gv, err := schema.ParseGroupVersion(resList.GroupVersion)
+		if err != nil {
+			gv = schema.GroupVersion{}
+		}
+		for _, resource := range resList.APIResources {
+			groupeResourceDef := schema.GroupVersionResource{Group: gv.Group, Version: gv.Version, Resource: resource.Name}
+			listResponse, err := dynamicClient.Resource(groupeResourceDef).Namespace(corev1.NamespaceAll).List(ctx, metav1.ListOptions{})
+			if err != nil {
+				if strings.Contains(err.Error(), errorFaildtoList) || strings.Contains(err.Error(), errorMethodNotAllowed) {
```

Comment:
Identifying errors in strings is not recommended.
You can convert the error to `StatusError` and determine the reason.

```go
statusErr := &apierrors.StatusError{}
if !errors.As(err, &statusErr) {
	return nil, err
}
if statusErr.ErrStatus.Reason == metav1.StatusReasonNotFound || statusErr.ErrStatus.Reason == metav1.StatusReasonMethodNotAllowed {
	continue
}
```

And add `import apierrors "k8s.io/apimachinery/pkg/api/errors"`